### ISSUE TITLE
[microprofile] The /openapi does not include resources from the service's other modules /CUSTCOM-104

### ISSUE THREAD

**Issue by lw-mcno** [payara/Payara#3769]
# Description #
----------

A MicroProfile service gets built using Gradle. It consists of a few Gradle modules, e.g. a myservice-core and myservice-war. The myservice-core is a dependency of the myservice-war module which gets packed into the .war archive as a JAR inside lib/. 

## Expected Outcome

@Path-annotated Resources from both the -core and -war modules are found during scanning and included in the output of the /openapi endpoint.

## Current Outcome

Only the Resources from the -war module are present in the /openapi output.

## Environment ##

- **Payara Version**: 5.184
- **Edition**: Micro
- **JDK Version**: 8 u181 - Oracle/IBM/OpenJDK -->
- **Operating System**: Windows

**Comment by lw-mcno** [comment]
I've created a small project that reproduces the behavior: [micro3769repro.zip](https://github.com/payara/Payara/files/2929781/micro3769repro.zip)

It consists of a `repro-parent` maven module, which contains 2 submodules:
* `repro-core`, which defines one own REST controller (`CoreHelloController`),
* `repro-war`, which is basically the output of MicroProfile starter (for MP 2.0), adjusted to be a submodule. This one also has a REST controller (`WarHelloController`). It dependens on `repro-core` and creates the .war file / bundle.

When the `repro-war` builds a .war file, the `repro-core` gets embedded in it as a .jar inside the .war's `lib/` directory.

After starting the generated bundle, both controllers are found and served:
```
'ROOT' REST Endpoints:
GET     /api/application.wadl
GET     /api/hello-core
GET     /api/hello-war
GET     /openapi/
GET     /openapi/application.wadl
```

Only the controller from `classes/` in the .war (`WarHelloController`) gets included in the `/openapi` output:

```
openapi: 3.0.0
info:
  title: OpenAPI scanning repro API
  contact:
    name: First Last
    email: contact@provider.com
  version: "1.0"
externalDocs:
  description: The bug description
  url: https://github.com/payara/Payara/issues/3769
servers:
- url: localhost:8080/api/
tags:
- name: openapi
- name: bug
- name: repro
paths:
  /api/hello-war:
    get:
      description: get greeted by the .war's controller
      operationId: sayHello
      responses:
        default:
          description: Default Response.
          content:
            '*/*':
              schema:
                type: string
components: {}
```

**Comment by charleech** [comment]
Hi,

I also have faced the same issue and have a chance to create a reproducible test case at

https://github.com/charleech/payara-evaluation/tree/master/payara-evaluation-openapi

I hope this may be useful.

**Comment by jbee** [comment]
Is this tested with the latest release where we updated to newer Open API version?

**Comment by charleech** [comment]
I'm testing against the version 5.191.

**Comment by charleech** [comment]
I'm testing against the version 5.192, the issue still presents.

Do we have a chance to support rendering the resource from other modules(jar file)?

**Comment by jbee** [comment]
Thanks @charleech for testing with 5.192. I opened an internal issue for it: PAYARA-3886.

**Comment by jbee** [comment]
@charleech 

> Do we have a chance to support rendering the resource from other modules(jar file)?

As far as I can see from looking at https://github.com/payara/Payara/blob/master/appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/OpenApiService.java#L221 the issue is that we only consider classes in `"WEB-INF/classes/"` of the top-most archive (the deployed application). It looks like archives within aren't considered.
So I'd say currently your inly change to get it rendered is to include the classes in the top most jar file.

I am trying to push this issue into the upcoming iteration (starting July). Maybe you can test modifying the method pointed out and see if adding some form or recursive call for archives does resolve your issue?

**Comment by rdebusscher** [comment]
Tracked by `CUSTCOM-104`

**Comment by rsoika** [comment]
I this issue still tracked? I still have the problem with  payara/server-full:5.201 that jax-rs resources placed in a lib/jar are not detected by openapi.

**Comment by rdebusscher** [comment]
Hi Ralph,

Yes, we have decided to rewrite part of the OpenAPI implementation within Payara. I cannot give you any estimation yet of when it will be available.

**Comment by rsoika** [comment]
Thanks for the feedback. But why did you at Payara implement OpenAPI by your self and did not use an implementation like https://github.com/smallrye/smallrye-open-api  
As far as I understand now it should not be a problem of payara?

**Comment by rdebusscher** [comment]
We are not using the SmallRye implementations so that we do not have the additional dependencies of the SmallRye projects which duplicates the frameworks we already have and so that we can write it against the internal APIs and not the standard API so that we can make a better and tighter integration with Payara.

**Comment by smillidge** [comment]
Fixed by #4768
